### PR TITLE
 MEN-4401: Separate module loggers & improved deployment logging

### PR DESCRIPTION
**Pull request by NilsOlavKJohansen**
Changelog: Log messages come from a separate module logger, instead of the root logger.

Signed-off-by: Nils Olav Kvelvane Johansen <nils.olav@northern.tech>

### CODE REVIEW

**Comment by oleorhagen** [comment]
I had a quick lookover on the general direction and structure of this, and it looks really promising! Just about what I had in mind, and I'm sure you also see that the logging were in need of some cleanup :)

I did not did too deep though, due to the amount of red pipes. Have you had a look at this already, or do you need help looking into this? @NilsOlavKJohansen

**Comment by olehermanse** [comment]
Seperat -> Separate

File: src/mender/client/authorize.py [hunk 1]
```diff
@@ -12,14 +12,17 @@
...
-import requests
 
 from cryptography.hazmat.primitives.asymmetric.rsa import RSAPrivateKeyWithSerialization
+import requests
 
 import mender.security.key as key
 
+log = logging.getLogger(__name__)
+
+
```
Comment (oleorhagen) [inline src/mender/client/authorize.py:25]: Question:
Why two newlines here?

File: src/mender/client/deployments.py [hunk 2]
```diff
@@ -11,19 +11,21 @@
...
 import re
-
 import requests
 
 import mender.settings.settings as settings
 import mender.log.log as menderlog
 from mender.client import HTTPUnathorized
 
+log = logging.getLogger(__name__)
+
```
Comment (oleorhagen) [inline src/mender/client/deployments.py:28]: And one here?

File: src/mender/statemachine/statemachine.py [hunk 3]
```diff
@@ -40,6 +40,8 @@ def __init__(self):
         self.config = config.Config({}, {})
         self.identity_data = {}
 
+    identity_data = None
```
Comment (oleorhagen) [inline src/mender/statemachine/statemachine.py:43]: Question:
What is this for?  `identity_data` is already an instance variable

File: src/mender/log/rootlogger.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+
+# Only used for testing, when mender.py has not been run.
+class SetUpLogForTesting:
+    log_file = False
+    log_level = "debug"
+    no_syslog = False
+
+
+def setup_logger(args=SetUpLogForTesting()):
```
Comment (oleorhagen) [inline src/mender/log/rootlogger.py:27]: Seems the `global settings` are not respected anylonger:
Try this command:
```
./.venv/bin/mender-python-client --data $(pwd) show-artifact
```
With the `var/lib/mender` folder existing locally in `PWD`

Comment (oleorhagen) [reply src/mender/log/rootlogger.py:27]: Why do you think this is? :)

File: src/mender/log/rootlogger.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+import logging
+from mender.log.log import DeploymentLogHandler
+
+log = logging.getLogger(__name__)
+
+
+# Only used for testing, when mender.py has not been run.
+class SetUpLogForTesting:
```
Comment (oleorhagen) [inline src/mender/log/rootlogger.py:21]: Question, why not pass this as a part of the test setup in `test_log.py`?

File: src/mender/mender.py [hunk 6]
```diff
@@ -215,9 +184,7 @@ def main(testargs=None):
         "--version", "-v", help="print the version", default=False, action="store_true"
     )
     args = parser.parse_args(testargs)
-    log.basicConfig(
-        datefmt="%Y-%m-%d %H:%M:%S", format="%(asctime)s %(levelname)-8s %(message)s",
-    )
+    rootlogger.setup_logger(args)
```
Comment (oleorhagen) [inline src/mender/mender.py:187]: Suggestion:
Isn't `rootlogger.setup()` a nicer interface. We already know it is a logger :)

File: src/mender/mender.py [hunk 7]
```diff
@@ -90,23 +90,14 @@ def report(args):
...
-            handler
-            for handler in log.getLogger("").handlers
-            if isinstance(handler, DeploymentLogHandler)
-        ]
-        assert (
-            len(update_handlers) == 1
-        ), "Something is wrong with the setup of the DeploymentLogHandler"
-        deployment_log_handler = update_handlers[0]
-        deployment_log_handler.enable()
+        log.parent.deployment_log_handler.enable()
```
Comment (oleorhagen) [inline src/mender/mender.py:93]: Question:
Can this not be handled from the `rootlogger`? 
On the surface, it just seems a little odd to import the `rootlogger`, and then call `log.parent` to receive something that is in the rootlogger (Or at least I would expect it to, being the rootlogger and all).

File: src/mender/log/rootlogger.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    log_level = "debug"
+    no_syslog = False
+
+
+def setup_logger(args=SetUpLogForTesting()):
+    logging.basicConfig(
+        datefmt="%Y-%m-%d %H:%M:%S",
+        format="%(name)s %(asctime)s %(levelname)-8s %(message)s",
+    )
+    root_logger = logging.getLogger("mender")
```
Comment (oleorhagen) [inline src/mender/log/rootlogger.py:32]: Hmm, I'm not fully understanding the logger hierarchy here. 
When you call:
```
root_logger = logging.getLogger("mender")
```
This inherits from the `rootLogger` no? :thinking:

**Review by oleorhagen — COMMENTED** [review]
I've left some comments and notes for further discussion.

This is definitely a step in the right direction, but I do think there is still some room for improvement here. I've left some comments we can discuss.

I also see that you've accidentally broken the `--data` flag, which is no longer respected, and in fact the client crashes when using it. 

Maybe this is a good time to actually add a regression test for this, so that we don't have to manually guard this in the future? :)

File: src/mender/log/log.py [hunk 9]
```diff
@@ -44,19 +45,16 @@ def format(self, record) -> str:
...
-
-    def handle(self, record):
-        if self.enabled:
-            super().handle(record)
+        self._reset()
+        self.disable()
 
     def enable(self, reset=False):
-        self.enabled = True
+        self.setLevel(logging.DEBUG)
```
[the remark this answers is not in the document]
Comment (NilsOlavKJohansen) [reply src/mender/log/log.py:57]: That was the idea, since we use the logger itself to set the level of messages to be logged. Therefore, all messages sent to the deployment handler should be logged. But if we want to change the log level of individual handlars then this could be a weakness.

File: src/mender/client/deployments.py (hunk 2 above)
Comment (NilsOlavKJohansen) [reply src/mender/client/deployments.py:28]: Oh one and two lines, that is an inconsistency. I have to fix it :)

File: src/mender/statemachine/statemachine.py (hunk 3 above)
Comment (NilsOlavKJohansen) [reply src/mender/statemachine/statemachine.py:43]: I did not see it either. The new one is going away.

File: src/mender/mender.py (hunk 6 above)
Comment (NilsOlavKJohansen) [reply src/mender/mender.py:187]: Yes, agreed : )

File: src/mender/log/rootlogger.py (hunk 5 above)
Comment (NilsOlavKJohansen) [reply src/mender/log/rootlogger.py:21]: The idea was that other tests such as test_mender and test_statementmachine could easily set up rootlogger and deploymentloghandler without having to use any arguments.
Just one call on `rootlogger.setup_logger()`

But maybe it's better to create the deployment log handler with monkeypatching in the unit tests?

File: src/mender/log/rootlogger.py (hunk 8 above)
Comment (NilsOlavKJohansen) [reply src/mender/log/rootlogger.py:32]: That's right, getLogger ("mender") it inherits from root. And all other loggers inherit from the "mender" logger.

It would have been nicer not to use logging.basicConfig and rather set up the handler and formats directly in the "mender" logger. Then there is no need to manipulate the "root" logger. It had also worked not to create the "mender" logger and only worked directly on the root. But my impression is that it is not customary.

Logger hierarchy:
mender logging

After drawing this diagram, I realize that the name rootLooger.py should be changed.

File: src/mender/log/rootlogger.py (hunk 4 above)
Comment (NilsOlavKJohansen) [reply src/mender/log/rootlogger.py:27]: `settings.PATHS = settings.Path(data_store=args.data)`

Hmmm, maybe it was not a great idea to create the deployment log handler before updating the data path :thinking:

But it would have been great to learn how to write regression tests : )

Comment (NilsOlavKJohansen) [reply src/mender/log/rootlogger.py:27]: Hmmm, maybe it's not a great idea to create the deployment log handler before updating the data path. :thinking: 

Line 187 ` rootlogger.setup_logger(args)`
Line 193 `settings.PATHS = settings.Path(data_store=args.data)`

But it would have been great to learn how to write regression tests. Maybe this could be the first one :)

Comment (oleorhagen) [reply src/mender/log/rootlogger.py]: A regression test is just a regular test, only for this one thing we just found, which used to work, but now broke, without being caught by any of the existing tests.
Thus testing that we never again introduce this specific regression :)

File: src/mender/log/rootlogger.py (hunk 5 above)
Comment (oleorhagen) [reply src/mender/log/rootlogger.py:21]: Hmm, I don't see why monkey-patching would be needed either.
Like I said, could it not be passed in as a part of the test setups?

File: src/mender/log/rootlogger.py (hunk 8 above)
Comment (oleorhagen) [reply src/mender/log/rootlogger.py]: > But my impression is that it is not customary
I have no idea to be honest. But if it server our cause here, I don't see why not :thinking: 

Following on this:
> It would have been nicer not to use logging.basicConfig and rather set up the handler and formats directly in the "mender" logger. Then there is no need to manipulate the "root" logger. 
I agree, and this was kinda what I thought would be the best approach when creating the ticket, mind you I did not have much oversight over the options at the time though.

But it does sound appealing, right?

File: src/mender/log/rootlogger.py (hunk 5 above)
Comment (NilsOlavKJohansen) [reply src/mender/log/rootlogger.py:21]: Yes, it will be tidier! I change the rootlogger and the tests :+1:

File: src/mender/mender.py (hunk 7 above)
Comment (NilsOlavKJohansen) [reply src/mender/mender.py:93]: Yes, I think it should work fine. `log.parent` corresponds to `logging.getLogger("mender")`. So it can be easily changed. Or do you think it is better to do it with a method in rootLogger.py?

File: src/mender/log/rootlogger.py (hunk 4 above)
Comment (NilsOlavKJohansen) [reply src/mender/log/rootlogger.py]: Good to know : )

File: src/mender/log/rootlogger.py (hunk 8 above)
Comment (NilsOlavKJohansen) [reply src/mender/log/rootlogger.py]: In the latest change, the "root" logger is just the default logging object.

All custom changes are made in the "mender" logger. This logs nothing itself, but all other loggers inherit its properties.

and then we never get into a situation where we have to write `log.getLogger("")` to get the root logger. But maybe it's just me who thinks it looks incredibly ugly :smile: 

But but if we change the line `log.getLogger("mender")` to `log.getLogger("")` we will make the changes directly on the root logger. Do you think it seems tidier?

File: src/mender/bootstrap/bootstrap.py [hunk 10]
```diff
@@ -12,15 +12,16 @@
...
 import logging as log
 import os.path
 from typing import Optional
-
 from cryptography.hazmat.primitives.asymmetric.rsa import RSAPrivateKeyWithSerialization
 from cryptography.exceptions import UnsupportedAlgorithm
-
 import mender.security.key as key
 
+log = logging.getLogger(__name__)
```
Comment (oleorhagen) [inline src/mender/bootstrap/bootstrap.py:23]: Question:
here, `logging` is imported as `log`, but you are referring to the import as `logging` still. Is this just a mistake?

Also, what's with the empty line changes?

File: tests/unit/test_mender.py [hunk 11]
```diff
@@ -129,7 +136,7 @@ def test_report(args, ctx, caplog, tmpdir, monkeypatch):
         m.setattr(statemachine, "Context", lambda *args, **kwargs: ctx)
         m.setattr(authorize, "request", lambda *args, **kwargs: "JWTToken")
         m.setattr(args, "failure", True)
-        with pytest.raises(AssertionError):
+        with pytest.raises(SystemExit):
```
Comment (oleorhagen) [inline tests/unit/test_mender.py:139]: Question:
Why was this changed?

File: tests/unit/test_statemachine.py [hunk 12]
```diff
@@ -109,6 +115,7 @@ def test_sync_inventory(self, ctx, monkeypatch, caplog):
             m.setattr(inventory, "aggregate", lambda *args, **kwargs: None)
             m.setattr(client_inventory, "request", lambda *args, **kwargs: None)
             sync_inventory.run(ctx)
+
```
Comment (oleorhagen) [inline tests/unit/test_statemachine.py:118]: Random newline

File: tests/unit/test_statemachine.py [hunk 13]
```diff
@@ -188,29 +195,23 @@ def test_unsupported_states(self, ctx):
 
 
 class TestStateMachines:
+    @pytest.fixture(autouse=True)
+    def setup_root_logger(self):
+        """Set the root logger and deployment handlar"""
```
Comment (oleorhagen) [inline tests/unit/test_statemachine.py:200]: `s/handlar/handler/`

File: tests/unit/test_log.py [hunk 14]
```diff
@@ -12,37 +12,38 @@
...
-import logging as testlogger
+import logging
 import os.path
 import pytest
-
-import mender.log.log as mlog
+import mender.log.rootlogger as rootlogger
 import mender.settings.settings as settings
 
+log = logging.getLogger("mender.logTesting")
```
Comment (oleorhagen) [inline tests/unit/test_log.py:21]: Question:
Why is this needed?

File: tests/unit/test_log.py [hunk 15]
```diff
@@ -12,37 +12,38 @@
...
-        logger = mlog.DeploymentLogHandler()
-        rootlogger = testlogger.getLogger("")
-        rootlogger.addHandler(logger)
-        return logger, log_file
+@pytest.fixture(name="set_up_rootlogger", autouse=True)
+def fixture_set_up_rootlogger(tmpdir):
+    d = tmpdir.mkdir("test_logger")
+    log_file = os.path.join(d, "deployment.log")
+    settings.PATHS.deployment_log = d
+    rootlogger.setup()
```
Comment (oleorhagen) [inline tests/unit/test_log.py:29]: Question:
Why don't you pass in the `test configuration` here? Instead of doing it in the production code?

File: src/mender/mender.py (hunk 7 above)
Comment (oleorhagen) [reply src/mender/mender.py:93]: Hmm, I think it would be nice to wrap it,
Like doesn't:
```
log.deployment_log_handler.enable()
```
Look better than:
```
log.parent.deployment_log_handler.enable()
```
?
But it is not critical of course

File: src/mender/bootstrap/bootstrap.py (hunk 10 above)
Comment (NilsOlavKJohansen) [reply src/mender/bootstrap/bootstrap.py:23]: Yea its a mistake, fixed it in my local repo already :+1:

File: tests/unit/test_log.py (hunk 14 above)
Comment (NilsOlavKJohansen) [reply tests/unit/test_log.py:21]: It allows the "log" to inherit the properties of the rootlogger. Otherwise these lines would not have worked:
`log.parent.deployment_log_handler.enable()`
`data = log.parent.deployment_log_handler.marshal()`

File: tests/unit/test_log.py (hunk 15 above)
Comment (NilsOlavKJohansen) [reply tests/unit/test_log.py:29]: The idea was to make it easier to set up the rootlogger. As several of the tests need to set it up. But agree that it is stupid to have it in the production code.  Now i do this in test_mender, test_statementmachine and test_log.

```

  class Args:
        log_file = False
        log_level = "debug"
        no_syslog = False

  rootlogger.setup(Args())
```

File: tests/unit/test_mender.py (hunk 11 above)
Comment (NilsOlavKJohansen) [reply tests/unit/test_mender.py:139]: To enable the deployment log handler.

Before we did this:
```
update_handlers = [
            handler
            for handler in log.getLogger("").handlers
            if isinstance(handler, DeploymentLogHandler)
        ]
assert (
            len(update_handlers) == 1
        ), "Something is wrong with the setup of the DeploymentLogHandler"
        deployment_log_handler = update_handlers[0]
deployment_log_handler = update_handlers[0]
deployment_log_handler.enable()

```
But now it is not possible that there is no deployment log handler as long as the root logger has been created. Therefore we can enable it by doing this:
```
log.parent.deployment_log_handler.enable()
```

Therefore, I had to remove the assertion error from the test.

```
    elif args.failure:
        log.info("Reporting a failed update to the Mender server")
        log.parent.deployment_log_handler.enable()
        if not deployments.report(
            context.config.ServerURL,
            deployments.STATUS_FAILURE,
            deployment_id,
            context.config.ServerCertificate,
            jwt,
            log.parent.deployment_log_handler,
        ):
            log.error("Failed to report the failed update status to the Mender server")
            sys.exit(1)
```
But since the assertion error is no longer there, the program continues to run, and deployments.report fails. and sys.exit (1) happens instead.

But maybe this should arrange in a different way? This test was actually used to check what happened when there were no deployment log handler. But now it checks what happens when:
`if not deployments.report()`

File: src/mender/mender.py (hunk 7 above)
Comment (NilsOlavKJohansen) [reply src/mender/mender.py:93]: I have tried it today, and unfortunately I do not manage to remove the "parent" in any good way : /

File: src/mender/log/rootlogger.py (hunk 8 above)
Comment (oleorhagen) [reply src/mender/log/rootlogger.py]: Hmm, I'm kinda undecided. You can choose :)

Comment (NilsOlavKJohansen) [reply src/mender/log/rootlogger.py]: Then I will keep using the "mender" logger. I think it's tidier : )

File: tests/unit/test_mender.py (hunk 11 above)
Comment (oleorhagen) [reply tests/unit/test_mender.py:145]: >But maybe this should arrange in a different way? This test was actually used to check what happened when there were no deployment log handler. But now it checks what happens when:
if not deployments.report()
Yes, exactly, and this is handled elsewhere I seem to recall :)
So, is there any way in which the `log.parent` call can fail? If not I guess testing it is not strictly necessary anylonger

File: src/mender/bootstrap/bootstrap.py [hunk 16]
```diff
@@ -12,15 +12,16 @@
 #    See the License for the specific language governing permissions and
 #    limitations under the License.
 
-import logging as log
 import os.path
+import logging
 from typing import Optional
-
```
Comment (oleorhagen) [inline src/mender/bootstrap/bootstrap.py:18]: Random whitespace removed

File: src/mender/client/authorize.py [hunk 17]
```diff
@@ -12,14 +12,17 @@
 #    See the License for the specific language governing permissions and
 #    limitations under the License.
 import json
-import logging as log
+import logging
 from typing import Optional
-import requests
```
Comment (oleorhagen) [inline src/mender/client/authorize.py:17]: Moved, why?

File: src/mender/client/deployments.py [hunk 18]
```diff
@@ -11,19 +11,22 @@
...
 #    limitations under the License.
-import logging as log
 from typing import Optional
 from typing import Dict
+import logging
+
 import os.path
 import time
 import re
-
```
Comment (oleorhagen) [inline src/mender/client/deployments.py:20]: Random whitespace removed

File: src/mender/log/log.py [hunk 19]
```diff
@@ -12,14 +12,15 @@
 #    See the License for the specific language governing permissions and
 #    limitations under the License.
 import json
-import logging as log
+import logging
 import logging.handlers
 import os
 import os.path
 from typing import List
-
```
Comment (oleorhagen) [inline src/mender/log/log.py:20]: Random whitespace removal

File: src/mender/log/rootlogger.py [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+        logging.NullHandler() if args.no_syslog else logging.handlers.SysLogHandler()
+    )
+    handlers.append(syslogger)
+    if args.log_file:
+        handlers.append(logging.FileHandler(args.log_file))
+
+    deployment_log_handler = DeploymentLogHandler()
+    handlers.append(deployment_log_handler)
+
+    root_logger = logging.getLogger("mender")
```
Comment (oleorhagen) [inline src/mender/log/rootlogger.py:47]: Musing:
Since this is really not the rootlogger, although it is sort of 'our' rootlogger. Maybe it should be named like `mender_logger` or something similar?
I'm not sure if this is better or not to be honest

File: src/mender/mender.py [hunk 21]
```diff
@@ -12,17 +12,17 @@
 #    See the License for the specific language governing permissions and
 #    limitations under the License.
 import argparse
-import logging as log
+import logging
 import sys
-
```
Comment (oleorhagen) [inline src/mender/mender.py:17]: Random whitespace removal

File: src/mender/mender.py [hunk 22]
```diff
@@ -12,17 +12,17 @@
...
 import mender.bootstrap.bootstrap as bootstrap
 import mender.client.authorize as authorize
 import mender.client.deployments as deployments
 import mender.settings.settings as settings
 import mender.statemachine.statemachine as statemachine
-
```
Comment (oleorhagen) [inline src/mender/mender.py:23]: same

File: src/mender/statemachine/statemachine.py [hunk 23]
```diff
@@ -11,11 +11,10 @@
 #    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 #    See the License for the specific language governing permissions and
 #    limitations under the License.
-import logging as log
+import logging
 import os.path
 import sys
 import time
-
```
Comment (oleorhagen) [inline src/mender/statemachine/statemachine.py:18]: Random whitespace removal

File: src/mender/statemachine/statemachine.py [hunk 24]
```diff
@@ -108,16 +108,7 @@ def __init__(self, force_bootstrap=False):
...
-            handler
-            for handler in log.getLogger("").handlers
-            if isinstance(handler, DeploymentLogHandler)
-        ]
-        assert (
-            len(deployment_log_handler) == 1
-        ), "Something is wrong with the setup of the DeploymentLogHandler"
-        self.context.deployment_log_handler = deployment_log_handler[0]
-        self.context.deployment_log_handler.disable()
+        log.parent.deployment_log_handler.disable()
```
Comment (oleorhagen) [inline src/mender/statemachine/statemachine.py:111]: So much better :tada: :raised_hands:

File: tests/unit/test_log.py [hunk 25]
```diff
@@ -12,37 +12,44 @@
 #    See the License for the specific language governing permissions and
 #    limitations under the License.
 import json
-import logging as testlogger
+import logging
 import os.path
 import pytest
-
```
Comment (oleorhagen) [inline tests/unit/test_log.py:18]: Whitespace removed

File: tests/unit/test_log.py [hunk 26]
```diff
@@ -12,37 +12,44 @@
...
-        rootlogger = testlogger.getLogger("")
-        rootlogger.addHandler(logger)
-        return logger, log_file
+@pytest.fixture(name="set_up_rootlogger", autouse=True)
+def fixture_set_up_rootlogger(tmpdir):
+    d = tmpdir.mkdir("test_logger")
+    log_file = os.path.join(d, "deployment.log")
+    settings.PATHS.deployment_log = d
+
+    class Args:
```
Comment (oleorhagen) [inline tests/unit/test_log.py:30]: :100:

File: tests/unit/test_log.py [hunk 27]
```diff
@@ -51,14 +58,11 @@ def test_logger(self, caplog, deployment_logger):
...
-    def test_log_marshalling(self, caplog, deployment_logger):
-        logger, _ = deployment_logger
-        logger.enable()
-        rootlogger = testlogger.getLogger("")
-        rootlogger.addHandler(logger)
-        testlogger.info("foo")
-        testlogger.error("bar")
-        data = logger.marshal()
+    def test_log_marshalling(self, caplog):
+        log.parent.deployment_log_handler.enable()
```
Comment (oleorhagen) [inline tests/unit/test_log.py:62]: :rocket:

**Review by oleorhagen — COMMENTED** [review]
I think it's coming together real nice!

File: src/mender/log/rootlogger.py (hunk 20 above)
Comment (NilsOlavKJohansen) [reply src/mender/log/rootlogger.py:47]: Yes, it would be less confusing :+1:

File: src/mender/client/authorize.py (hunk 17 above)
Comment (NilsOlavKJohansen) [reply src/mender/client/authorize.py:17]: Oh, sorry, missed that one. I will fix it right now :+1:

Comment (oleorhagen) [reply src/mender/client/authorize.py:17]: https://github.com/mendersoftware/mender-python-client/pull/64 Once this is merged, you can simply rebase, and we don't have to worry about imports anymore :raised_hands:

File: tests/unit/test_mender.py (hunk 11 above)
Comment (NilsOlavKJohansen) [reply tests/unit/test_mender.py:145]: Log.parent may fail IF the root logger is not set up. It is set up by the call `rootlogger.setup() `in mender.py

But maybe it's a good idea to write a test in test_mender that checks that `rootlogger.setup()`is called?

Comment (oleorhagen) [reply tests/unit/test_mender.py:145]: Yes, this sounds like a good idea to me :)

File: src/mender/log/log.py [hunk 28]
```diff
@@ -44,19 +46,18 @@ def format(self, record) -> str:
...
-    def handle(self, record):
-        if self.enabled:
-            super().handle(record)
+        self._reset()
+        self.disable()
 
     def enable(self, reset=False):
-        self.enabled = True
+        # Sets the log level to 0.
+        # All messages that the logger emits to the deployment log handler will be logged.
```
Comment (oleorhagen) [inline src/mender/log/log.py:59]: :tada:

File: src/mender/log/rootlogger.py [hunk 29]
```diff
@@ -0,0 +1,52 @@
+# Copyright 2021 Northern.tech AS
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (oleorhagen) [inline src/mender/log/rootlogger.py:3]: File is still named `rootlogger` Maybe it should mirror the new name?

**Review by oleorhagen — COMMENTED** [review]
Just one thing left we should discuss :)

File: src/mender/log/rootlogger.py (hunk 29 above)
Comment (NilsOlavKJohansen) [reply src/mender/log/rootlogger.py:3]: Good idea!

**Review by oleorhagen — APPROVED** [review]
Good job!
